### ISSUE TITLE
if i do nms(agnostic=True), why compare with conf_thres twice?

### ISSUE THREAD

**Issue by Fred199683** [ultralytics/yolov5#3065]
## ❔Question
if i only have one class, the training process doesn't care about the Loss_cls(=0). And the output of the model can be like this:[xmin,ymin,xmax,ymax,obj_conf,cls_conf], so i think the last element ouput[5] is not trained, and it's not useful. But in your nms function: first of all you filter out the box which obj_conf is lower than threshold, but then u multiple obj_conf and cls_conf and compare with threshold again. I don't think it's right, because the cls_conf is not reliable, it's never trained.
can anyone help me with this?

## Additional context
`        else:  # best class only
            conf, j = x[:, 5:].max(1, keepdim=True)
            x = torch.cat((box, conf, j.float()), 1)[conf.view(-1) > conf_thres]`

`                if self.nc > 1:  # cls loss (only if multiple classes)
                    t = torch.full_like(ps[:, 5:], self.cn, device=device)  # targets
                    t[range(n), tcls[i]] = self.cp
                    lcls += self.BCEcls(ps[:, 5:], t)  # BCE`

**Comment by glenn-jocher** [comment]
@Fred199683 for single class use cases cls conf is initialized high enough to convert to a 1.0 value after a sigmoid operation. Since it is not modified the value remains after training.

If you have an alternative implementation please provide profiling results and test results with the default repo and with your proposed change if you would like to us to consider a modification to the codebase. Thank you!

**Comment by Fred199683** [comment]
> @Fred199683 for single class use cases cls conf is initialized high enough to convert to a 1.0 value after a sigmoid operation. Since it is not modified the value remains after training.
> 
> If you have an alternative implementation please provide profiling results and test results with the default repo and with your proposed change if you would like to us to consider a modification to the codebase. Thank you!

Can u show me where do u initialize the cls conf to almost 1.0? I can't find it....Thx!

**Comment by glenn-jocher** [comment]
@Fred199683 yolo.py contains code for initialization of output layers.

**Comment by glenn-jocher** [comment]
@Fred199683 you can find the output initialization here:
https://github.com/ultralytics/yolov5/blob/e97d129db40d3b3592fca7d1c8a6fe39133e7e47/models/yolo.py#L172-L181

**Comment by Fred199683** [comment]
> @Fred199683 you can find the output initialization here:
> https://github.com/ultralytics/yolov5/blob/e97d129db40d3b3592fca7d1c8a6fe39133e7e47/models/yolo.py#L172-L181
